File: openadmin/spec/markdown.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from typing import Literal, NotRequired, TypedDict

from .colors import Color
from .http_methods import HttpMethod
from .icons import Icon
from .json_schema import JsonSchema


class MarkdownComponent(TypedDict):
    type: Literal["markdown"]
    id: str
    name: str
    description: str | None
    color: Color | None
    icon: Icon | None
    refresh: int | None
    method: HttpMethod
    form: JsonSchema | None
    body: JsonSchema | None
    query: JsonSchema | None


type MarkdownContent = str


class MarkdownResponse(TypedDict):
    icon: NotRequired[Icon]
    color: NotRequired[Color]
    refresh: NotRequired[int | None]
    content: MarkdownContent


type Markdown = MarkdownContent | MarkdownResponse

File: client/src/schemas/markdown.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { colorSchema } from "./color"
import { httpMethodSchema } from "./http-method"
import { iconSchema } from "./icon"
import { jsonSchemaSchema } from "./json-schema"

export const markdownComponentSchema = z.object({
	type: z.literal("markdown"),
	id: z.string(),
	name: z.string(),
	description: z.string().nullable(),
	color: colorSchema.nullable(),
	icon: iconSchema.nullable(),
	refresh: z.number().nullable(),
	method: httpMethodSchema,
	form: jsonSchemaSchema.nullable(),
	body: jsonSchemaSchema.nullable(),
	query: jsonSchemaSchema.nullable(),
})

export type MarkdownComponent = z.infer<typeof markdownComponentSchema>

export const markdownContentSchema = z.string()

export type MarkdownContent = z.infer<typeof markdownContentSchema>

export const markdownResponseSchema = z.object({
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
	refresh: z.number().nullable().optional(),
	content: markdownContentSchema,
})

export type MarkdownResponse = z.infer<typeof markdownResponseSchema>

export const markdownSchema = z.union([markdownContentSchema, markdownResponseSchema])

export type Markdown = z.infer<typeof markdownSchema>
